src/twisted/internet/_resolver.py

```diff
@@ -169,6 +171,19 @@ def resolveHostName(self, resolutionReceiver, hostName, portNumber=0,
 
         @return: see interface
         """
+        # If it's str, we need to make sure that it's just ASCII.
+        try:
+            hostName = hostName.encode('ascii')
+        except UnicodeEncodeError:
+            # If it's not just ASCII, IDNA it. We don't want to give a Unicode
+            # string with non-ASCII in it to Python 3, as if anyone passes that
+            # to a Python 3 stdlib function, it will probably use the wrong
+            # IDNA version and break absolutely everything
+            hostName = _idnaBytes(hostName)
+
+        # Make sure it's passed down as a native str, to maintain the interface
+        hostName = nativeString(hostName)
```

Comment:
I think this is correct.  `SimpleResolverComplexifier` implements [`IHostnameResolver`](http://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IHostnameResolver.html), so its [`resolveHostname`](http://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IHostnameResolver.html#resolveHostName) method accepts `unicode` objects.

If `hostName = u'192.168.1.1'`, then by line 186 `hostName = '192.168.1.1'` on Python 2 and Python 3, so that `SimpleResolverComplexifier.resolveHostName` always receives a `str`.  The same is true for any pure-ASCII domain name.

If `hostName = u'шанти-шанти.рф'`, then by line 186 `hostName = 'xn----7sbb4ac0ad0be6cf.xn--p1ai'`.  This is also a native string.

Resolvers that call the system's resolver, like `ThreadedResolver`, will resolve these values even on Python 3.  As the comment states, Python 3 uses the built in `idna` codec to encode native strings to bytes, and that codec encodes codepoints that fit within a single byte as their ASCII equivalents.  That is, the `idna` codec is idempotent.

The net result is that the issue's reproducing script now works!